File: src/commands/docker.rs
use anyhow::{Result, bail};
use clap::Subcommand;

use crate::client::UnraidClient;
use crate::graphql::get_docker_containers::GetDockerContainersDockerContainers as Container;
use crate::graphql::{
    ContainerState, GetDockerContainers, StartDockerContainer, StopDockerContainer,
    UpdateDockerContainer,
};

#[derive(Debug, Subcommand)]
pub enum DockerCommands {
    /// List Docker containers
    #[command(visible_alias = "ls")]
    ListContainers {
        /// Show all containers (default: only running)
        #[arg(short, long)]
        all: bool,
    },
    /// Start a Docker container
    Start {
        /// Container name
        name: String,
    },
    /// Stop a Docker container
    Stop {
        /// Container name
        name: String,
    },
    /// Restart a Docker container (stop then start)
    Restart {
        /// Container name
        name: String,
    },
    /// Update a Docker container to the latest image
    Update {
        /// Container name
        name: String,
    },
}

pub async fn handle_docker_command(cmd: DockerCommands, client: &UnraidClient) -> Result<()> {
    match cmd {
        DockerCommands::ListContainers { all } => list_containers(client, all).await,
        DockerCommands::Start { name } => start_container(client, &name).await,
        DockerCommands::Stop { name } => stop_container(client, &name).await,
        DockerCommands::Restart { name } => restart_container(client, &name).await,
        DockerCommands::Update { name } => update_container(client, &name).await,
    }
}

/// Resolve a container name to its prefixed ID by querying all containers.
async fn resolve_container_id(client: &UnraidClient, name: &str) -> Result<String> {
    let response = client
        .execute::<GetDockerContainers>(crate::graphql::get_docker_containers::Variables {})
        .await?;

    find_container_id(&response.docker.containers, name)
}

fn find_container_id(containers: &[Container], name: &str) -> Result<String> {
    let name_lower = name.to_lowercase();

    for container in containers {
        for container_name in &container.names {
            let clean = container_name.trim_start_matches('/').to_lowercase();
            if clean == name_lower {
                return Ok(container.id.clone());
            }
        }
    }

    bail!(
        "Container '{name}' not found. Use 'docker list-containers --all' to see available containers."
    );
}

async fn start_container(client: &UnraidClient, name: &str) -> Result<()> {
    let id = resolve_container_id(client, name).await?;

    println!("Starting container '{name}'...");
    let response = client
        .execute::<StartDockerContainer>(crate::graphql::start_docker_container::Variables { id })
        .await?;

    let container = response.docker.start;
    let state = format!("{:?}", container.state).to_lowercase();
    println!("Container '{name}' is now {state}.");

    Ok(())
}

async fn stop_container(client: &UnraidClient, name: &str) -> Result<()> {
    let id = resolve_container_id(client, name).await?;

    println!("Stopping container '{name}'...");
    let response = client
        .execute::<StopDockerContainer>(crate::graphql::stop_docker_container::Variables { id })
        .await?;

    let container = response.docker.stop;
    let state = format!("{:?}", container.state).to_lowercase();
    println!("Container '{name}' is now {state}.");

    Ok(())
}

async fn update_container(client: &UnraidClient, name: &str) -> Result<()> {
    let id = resolve_container_id(client, name).await?;

    println!("Updating container '{name}'...");
    let response = client
        .execute::<UpdateDockerContainer>(crate::graphql::update_docker_container::Variables { id })
        .await?;

    let container = response.docker.update_container;
    let state = format!("{:?}", container.state).to_lowercase();
    println!("Container '{name}' updated successfully (state: {state}).");

    Ok(())
}

async fn restart_container(client: &UnraidClient, name: &str) -> Result<()> {
    let id = resolve_container_id(client, name).await?;

    println!("Restarting container '{name}'...");
    client
        .execute::<StopDockerContainer>(crate::graphql::stop_docker_container::Variables {
            id: id.clone(),
        })
        .await?;

    let response = client
        .execute::<StartDockerContainer>(crate::graphql::start_docker_container::Variables { id })
        .await?;

    let container = response.docker.start;
    let state = format!("{:?}", container.state).to_lowercase();
    println!("Container '{name}' is now {state}.");

    Ok(())
}

async fn list_containers(client: &UnraidClient, show_all: bool) -> Result<()> {
    let response = client
        .execute::<GetDockerContainers>(crate::graphql::get_docker_containers::Variables {})
        .await?;

    let containers = response.docker.containers;
    let filtered = filter_by_state(containers, show_all);

    if filtered.is_empty() {
        if show_all {
            println!("No containers found.");
        } else {
            println!("No running containers found. Use --all to show all containers.");
        }
        return Ok(());
    }

    // Print header
    println!(
        "{:<30} {:<40} {:<10} {:<20}",
        "NAME", "IMAGE", "STATE", "STATUS"
    );
    println!("{}", "-".repeat(100));

    // Print containers
    for container in filtered {
        let name = container
            .names
            .first()
            .map_or("unnamed", |s| s.trim_start_matches('/'));

        let state = match container.state {
            ContainerState::RUNNING => "running",
            ContainerState::PAUSED => "paused",
            ContainerState::EXITED => "exited",
            ContainerState::Other(ref s) => s.as_str(),
        };

        println!(
            "{:<30} {:<40} {:<10} {:<20}",
            truncate(name, 29),
            truncate(&container.image, 39),
            state,
            truncate(&container.status, 19)
        );
    }

    Ok(())
}

fn filter_by_state(containers: Vec<Container>, show_all: bool) -> Vec<Container> {
    if show_all {
        containers
    } else {
        containers
            .into_iter()
            .filter(|c| c.state == ContainerState::RUNNING)
            .collect()
    }
}

fn truncate(s: &str, max_len: usize) -> String {
    if s.len() <= max_len {
        s.to_string()
    } else {
        format!("{}...", &s[..max_len - 3])
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    fn sample_container(id: &str, name: &str, state: ContainerState) -> Container {
        Container {
            id: id.to_string(),
            names: vec![name.to_string()],
            image: "some-image:latest".to_string(),
            state,
            status: "Up 2 hours".to_string(),
            ports: vec![],
        }
    }

    fn sample_containers() -> Vec<Container> {
        vec![
            sample_container("id-1", "plex", ContainerState::RUNNING),
            sample_container("id-2", "sonarr", ContainerState::RUNNING),
            sample_container("id-3", "radarr", ContainerState::EXITED),
            sample_container("id-4", "nginx", ContainerState::PAUSED),
        ]
    }

    #[test]
    fn truncate_truncates_correctly() {
        let cases = [
            ("hello", 10, "hello"),
            ("hello", 5, "hello"),
            ("hello world", 8, "hello..."),
            ("", 5, ""),
        ];

        for (input, max_len, expected) in cases {
            assert_eq!(
                truncate(input, max_len),
                expected,
                "truncate({input:?}, {max_len})"
            );
        }
    }

    // find_container_id tests

    #[test]
    fn find_container_id_returns_id_for_matching_name() {
        let containers = sample_containers();
        let result = find_container_id(&containers, "plex").unwrap();
        assert_eq!(result, "id-1");
    }

    #[test]
    fn find_container_id_is_case_insensitive() {
        let containers = sample_containers();
        let result = find_container_id(&containers, "PLEX").unwrap();
        assert_eq!(result, "id-1");
    }

    #[test]
    fn find_container_id_strips_leading_slash() {
        let containers = vec![Container {
            id: "id-slash".to_string(),
            names: vec!["/mycontainer".to_string()],
            image: "img".to_string(),
            state: ContainerState::RUNNING,
            status: "Up".to_string(),
            ports: vec![],
        }];

        let result = find_container_id(&containers, "mycontainer").unwrap();
        assert_eq!(result, "id-slash");
    }

    #[test]
    fn find_container_id_returns_error_for_unknown_name() {
        let containers = sample_containers();
        let err = find_container_id(&containers, "nonexistent").unwrap_err();
        assert!(err.to_string().contains("not found"));
    }

    #[test]
    fn find_container_id_returns_error_for_empty_list() {
        let err = find_container_id(&[], "anything").unwrap_err();
        assert!(err.to_string().contains("not found"));
    }

    #[test]
    fn find_container_id_matches_second_name_in_list() {
        let containers = vec![Container {
            id: "id-multi".to_string(),
            names: vec!["primary".to_string(), "alias".to_string()],
            image: "img".to_string(),
            state: ContainerState::RUNNING,
            status: "Up".to_string(),
            ports: vec![],
        }];

        let result = find_container_id(&containers, "alias").unwrap();
        assert_eq!(result, "id-multi");
    }

    // filter_by_state tests

    #[test]
    fn filter_by_state_returns_all_containers_when_show_all_is_true() {
        let containers = sample_containers();
        let filtered = filter_by_state(containers, true);
        assert_eq!(filtered.len(), 4);
    }

    #[test]
    fn filter_by_state_returns_only_running_when_show_all_is_false() {
        let containers = sample_containers();
        let filtered = filter_by_state(containers, false);
        assert_eq!(filtered.len(), 2);
        assert!(filtered.iter().all(|c| c.state == ContainerState::RUNNING));
    }

    #[test]
    fn filter_by_state_returns_empty_when_no_running_containers() {
        let containers = vec![
            sample_container("id-1", "a", ContainerState::EXITED),
            sample_container("id-2", "b", ContainerState::PAUSED),
        ];
        let filtered = filter_by_state(containers, false);
        assert!(filtered.is_empty());
    }
}

File: internal/commands/docker.go
package commands

import (
	"context"
	"fmt"
	"log"
	"strings"

	"github.com/Khan/genqlient/graphql"
	"github.com/spf13/cobra"

	"github.com/medzin/unraid-cli/internal/client"
)

type container = client.GetDockerContainersDockerContainersDockerContainer

func newDockerCmd(preRun func(*cobra.Command, []string) error) *cobra.Command {
	cmd := &cobra.Command{
		Use:   "docker",
		Short: "Docker container management",
	}

	cmd.AddCommand(
		newDockerListCmd(),
		newDockerLogsCmd(),
		newDockerStartCmd(),
		newDockerStopCmd(),
		newDockerRestartCmd(),
		newDockerPauseCmd(),
		newDockerUnpauseCmd(),
		newDockerUpdateCmd(),
	)
	for _, sub := range cmd.Commands() {
		sub.PreRunE = preRun
	}

	return cmd
}

func newDockerListCmd() *cobra.Command {
	var all bool

	cmd := &cobra.Command{
		Use:     "list",
		Aliases: []string{"ls"},
		Short:   "List Docker containers",
		Args:    cobra.NoArgs,
		RunE: func(cmd *cobra.Command, _ []string) error {
			return listContainers(cmd.Context(), getClient(cmd.Context()), all)
		},
	}

	cmd.Flags().BoolVarP(&all, "all", "a", false, "show all containers (default: only running)")
	return cmd
}

func newDockerStartCmd() *cobra.Command {
	return &cobra.Command{
		Use:   "start <name>",
		Short: "Start a Docker container",
		Args:  cobra.ExactArgs(1),
		RunE: func(cmd *cobra.Command, args []string) error {
			return startContainer(cmd.Context(), getClient(cmd.Context()), args[0])
		},
	}
}

func newDockerStopCmd() *cobra.Command {
	return &cobra.Command{
		Use:   "stop <name>",
		Short: "Stop a Docker container",
		Args:  cobra.ExactArgs(1),
		RunE: func(cmd *cobra.Command, args []string) error {
			return stopContainer(cmd.Context(), getClient(cmd.Context()), args[0])
		},
	}
}

func newDockerRestartCmd() *cobra.Command {
	return &cobra.Command{
		Use:   "restart <name>",
		Short: "Restart a Docker container (stop then start)",
		Args:  cobra.ExactArgs(1),
		RunE: func(cmd *cobra.Command, args []string) error {
			return restartContainer(cmd.Context(), getClient(cmd.Context()), args[0])
		},
	}
}

func newDockerPauseCmd() *cobra.Command {
	return &cobra.Command{
		Use:   "pause <name>",
		Short: "Pause a Docker container",
		Args:  cobra.ExactArgs(1),
		RunE: func(cmd *cobra.Command, args []string) error {
			return pauseContainer(cmd.Context(), getClient(cmd.Context()), args[0])
		},
	}
}

func newDockerUnpauseCmd() *cobra.Command {
	return &cobra.Command{
		Use:   "unpause <name>",
		Short: "Unpause a Docker container",
		Args:  cobra.ExactArgs(1),
		RunE: func(cmd *cobra.Command, args []string) error {
			return unpauseContainer(cmd.Context(), getClient(cmd.Context()), args[0])
		},
	}
}

func newDockerUpdateCmd() *cobra.Command {
	return &cobra.Command{
		Use:   "update <name>",
		Short: "Update a Docker container to the latest image",
		Args:  cobra.ExactArgs(1),
		RunE: func(cmd *cobra.Command, args []string) error {
			return updateContainer(cmd.Context(), getClient(cmd.Context()), args[0])
		},
	}
}

func newDockerLogsCmd() *cobra.Command {
	var tail int

	cmd := &cobra.Command{
		Use:   "logs <name>",
		Short: "Fetch container logs",
		Args:  cobra.ExactArgs(1),
		RunE: func(cmd *cobra.Command, args []string) error {
			return containerLogs(cmd.Context(), getClient(cmd.Context()), args[0], tail)
		},
	}

	cmd.Flags().IntVarP(&tail, "lines", "n", 100, "number of lines to show from the end")
	return cmd
}

func containerLogs(ctx context.Context, c graphql.Client, name string, tail int) error {
	id, err := resolveContainerID(ctx, c, name)
	if err != nil {
		return err
	}
	resp, err := client.GetContainerLogs(ctx, c, id, &tail)
	if err != nil {
		return err
	}
	lines := resp.Docker.Logs.Lines
	return render(ctx, lines, func() error {
		w := getOutputWriter(ctx)
		for _, line := range lines {
			if _, err := fmt.Fprintf(w, "%s  %s\n", line.Timestamp, line.Message); err != nil {
				return err
			}
		}
		return nil
	})
}

type containerMutationFunc func(ctx context.Context, c graphql.Client, id string) (string, error)

func containerAction(ctx context.Context, c graphql.Client, name, presentVerb string, mutation containerMutationFunc) error {
	id, err := resolveContainerID(ctx, c, name)
	if err != nil {
		return err
	}
	log.Printf("%s container '%s'...", presentVerb, name)
	state, err := mutation(ctx, c, id)
	if err != nil {
		return err
	}
	return printAction(ctx, fmt.Sprintf("Container '%s' is now %s.", name, state))
}

func resolveContainerID(ctx context.Context, c graphql.Client, name string) (string, error) {
	resp, err := client.GetDockerContainers(ctx, c)
	if err != nil {
		return "", err
	}
	return findContainerID(resp.Docker.Containers, name)
}

func findContainerID(containers []container, name string) (string, error) {
	nameLower := strings.ToLower(name)

	for _, ct := range containers {
		for _, cName := range ct.Names {
			clean := strings.ToLower(strings.TrimPrefix(cName, "/"))
			if clean == nameLower {
				return ct.Id, nil
			}
		}
	}

	return "", fmt.Errorf("container '%s' not found. Use 'docker list --all' to see available containers", name)
}

func startContainer(ctx context.Context, c graphql.Client, name string) error {
	return containerAction(ctx, c, name, "Starting", func(ctx context.Context, c graphql.Client, id string) (string, error) {
		resp, err := client.StartDockerContainer(ctx, c, id)
		if err != nil {
			return "", err
		}
		return strings.ToLower(string(resp.Docker.Start.State)), nil
	})
}

func stopContainer(ctx context.Context, c graphql.Client, name string) error {
	return containerAction(ctx, c, name, "Stopping", func(ctx context.Context, c graphql.Client, id string) (string, error) {
		resp, err := client.StopDockerContainer(ctx, c, id)
		if err != nil {
			return "", err
		}
		return strings.ToLower(string(resp.Docker.Stop.State)), nil
	})
}

func pauseContainer(ctx context.Context, c graphql.Client, name string) error {
	return containerAction(ctx, c, name, "Pausing", func(ctx context.Context, c graphql.Client, id string) (string, error) {
		resp, err := client.PauseDockerContainer(ctx, c, id)
		if err != nil {
			return "", err
		}
		return strings.ToLower(string(resp.Docker.Pause.State)), nil
	})
}

func unpauseContainer(ctx context.Context, c graphql.Client, name string) error {
	return containerAction(ctx, c, name, "Unpausing", func(ctx context.Context, c graphql.Client, id string) (string, error) {
		resp, err := client.UnpauseDockerContainer(ctx, c, id)
		if err != nil {
			return "", err
		}
		return strings.ToLower(string(resp.Docker.Unpause.State)), nil
	})
}

func restartContainer(ctx context.Context, c graphql.Client, name string) error {
	return containerAction(ctx, c, name, "Restarting", func(ctx context.Context, c graphql.Client, id string) (string, error) {
		if _, err := client.StopDockerContainer(ctx, c, id); err != nil {
			return "", err
		}
		resp, err := client.StartDockerContainer(ctx, c, id)
		if err != nil {
			return "", err
		}
		return strings.ToLower(string(resp.Docker.Start.State)), nil
	})
}

func updateContainer(ctx context.Context, c graphql.Client, name string) error {
	id, err := resolveContainerID(ctx, c, name)
	if err != nil {
		return err
	}

	log.Printf("Updating container '%s'...", name)
	resp, err := client.UpdateDockerContainer(ctx, c, id)
	if err != nil {
		return err
	}

	state := strings.ToLower(string(resp.Docker.UpdateContainer.State))
	return printAction(ctx, fmt.Sprintf("Container '%s' updated successfully (state: %s).", name, state))
}

func listContainers(ctx context.Context, c graphql.Client, showAll bool) error {
	resp, err := client.GetDockerContainers(ctx, c)
	if err != nil {
		return err
	}

	containers := filterContainersByState(resp.Docker.Containers, showAll)

	if containers == nil {
		containers = []container{}
	}

	return render(ctx, containers, func() error {
		w := getOutputWriter(ctx)
		if len(containers) == 0 {
			if showAll {
				_, err := fmt.Fprintln(w, "No containers found.")
				return err
			}
			_, err := fmt.Fprintln(w, "No running containers found. Use --all to show all containers.")
			return err
		}

		if _, err := fmt.Fprintf(w, "%-30s %-40s %-10s %-20s\n", "NAME", "IMAGE", "STATE", "STATUS"); err != nil {
			return err
		}
		if _, err := fmt.Fprintln(w, strings.Repeat("-", 100)); err != nil {
			return err
		}

		for _, ct := range containers {
			name := "unnamed"
			if len(ct.Names) > 0 {
				name = strings.TrimPrefix(ct.Names[0], "/")
			}
			if _, err := fmt.Fprintf(w, "%-30s %-40s %-10s %-20s\n",
				truncate(name, 29),
				truncate(ct.Image, 39),
				strings.ToLower(string(ct.State)),
				truncate(ct.Status, 19),
			); err != nil {
				return err
			}
		}

		return nil
	})
}

func filterContainersByState(containers []container, showAll bool) []container {
	if showAll {
		return containers
	}
	var filtered []container
	for _, ct := range containers {
		if ct.State == client.ContainerStateRunning {
			filtered = append(filtered, ct)
		}
	}
	return filtered
}
